h2/src/main/org/h2/mvstore/MVMap.java

```diff
@@ -1148,6 +1166,135 @@ private Page copy(Page source) {
+        assert split.map == this;
+        assert pos != null;
+        assert split.getKeyCount() > 0;
+        Object key = split.getKey(0);
+        assert pos.index < 0 : pos.index;
+        int index = -pos.index - 1;
+        assert index == pos.page.getKeyCount() : index + " != " + pos.page.getKeyCount();
+        Page p = pos.page;
+        pos = pos.parent;
+        CursorPos tip = pos;
+        int unsavedMemory = 0;
+        while (true) {
+            if (pos == null) {
+                if (p.getKeyCount() == 0) {
+                    p = split;
+                } else {
+                    Object keys[] = new Object[] { key };
+                    Page.PageReference children[] = new Page.PageReference[store.getKeysPerPage() + 1];
+                    children[0] = new Page.PageReference(p);
+                    children[1] = new Page.PageReference(split);
+                    p = Page.create(this, keys, null, children, p.getTotalCount() + split.getTotalCount(), 0);
+                }
+                break;
+            }
+            Page c = p;
+            p = pos.page;
+            index = pos.index;
+            pos = pos.parent;
+            p = p.copy();
+            p.setChild(index, split);
+            p.insertNode(index, key, c);
+            int keyCount;
+            if ((keyCount = p.getKeyCount()) <= store.getKeysPerPage() && (p.getMemory() < store.getMaxPageSize() || keyCount <= (p.isLeaf() ? 1 : 2))) {
+                break;
+            }
+            int at = keyCount - 2;
+            key = p.getKey(at);
+            split = p.split(at);
+            unsavedMemory += p.getMemory() + split.getMemory();
+        }
+        unsavedMemory += p.getMemory();
+        while (pos != null) {
+            Page c = p;
+            p = pos.page;
+            p = p.copy();
+            p.setChild(pos.index, c);
+            unsavedMemory += p.getMemory();
+            pos = pos.parent;
+        }
+        RootReference updatedRootReference = new RootReference(rootReference, p, ++attempt);
+        if(root.compareAndSet(rootReference, updatedRootReference)) {
+            while (tip != null) {
+                tip.page.removePage();
+                tip = tip.parent;
+            }
+            if (store.getFileStore() != null) {
+                store.registerUnsavedPage(unsavedMemory);
+            }
+            return updatedRootReference;
+        }
+        return null;
+    }
+
+    public void append(K key, V value) {
+        int attempt = 0;
+        boolean success = false;
+        while(!success) {
+            RootReference rootReference = getRoot();
+            if (rootReference.appendCounter >= keysPerPage) {
+                rootReference = flushAppendBuffer(rootReference);
+                assert rootReference.appendCounter < keysPerPage;
+            }
+            keysBuffer[rootReference.appendCounter] = key;
+            valuesBuffer[rootReference.appendCounter] = value;
```

Comment:
I don't see any protection against concurrent execution of `MVMap.append()` if MVStore is used by itself. `MVMap` implements `ConcurrentMap` so it's not expected that there are any public methods, including non-standard ones, that are not thread-safe.

With H2 at least session should be locked, however.

Reply:
This point is debatable. The fact that MVMap implements ConcurrentMap means absolutely nothing in relation to other methods, public or otherwise. Of course, this should be prominently mentioned in javadoc, which is due, my bad.
When used within H2, it is not necessary to hold any locks, and coincidental fact that session is locked, would not protect from multi-threaded abuse, although actual usage is single-threaded by design.